File: mappers-lua/mmc4.lua
MapperName = "MMC4"
MapperNumber = 10
DefaultPrgSize = 512 * 1024
DefaultChrSize = 256 * 1024

function WriteMMC4(address, data)
	WriteCpu(address, {data})
end

function DumpPrg(size)
	local banks = size / 0x4000
	for b = 0, banks-2 do
		print("Reading PRG bank #" .. tostring(b) .. "...")
		WriteMMC4(0xA000, b)
		ReadAddPrg(0x8000, 0x4000)
	end

	print("Reading last PRG bank #" .. tostring(banks-1) .. "...")
	ReadAddPrg(0xC000, 0x4000)
end

function DumpChr(size)
	local banks = size / 0x1000
	for b = 0, banks-1, 2 do
		print("Reading CHR banks #" .. tostring(b) .. ", #" .. tostring(b+1) .. "...")
		WriteMMC4(0xB000, b)
		WriteMMC4(0xC000, b)
		WriteMMC4(0xD000, b+1)
		WriteMMC4(0xE000, b+1)
		ReadAddChr(0x0000, 0x2000)
	end
end

function EnablePrgRam(size)

end

File: mappers-cs/MMC4.cs
using System;
using System.Collections.Generic;

namespace com.clusterrr.Famicom.Mappers
{
    public class MMC4 : IMapper
    {
        public string Name
        {
            get { return "MMC4"; }
        }

        public int Number
        {
            get { return 10; }
        }

        public string UnifName
        {
            get { return null; }
        }

        public int DefaultPrgSize
        {
            get { return 256 * 1024; }
        }

        public int DefaultChrSize
        {
            get { return 0; }
        }

        public void DumpPrg(FamicomDumperConnection dumper, List<byte> data, int size)
        {
            int banks = size / 0x4000;
            
            for (int bank = 0; bank < banks; bank++)
            {
                Console.Write("Reading PRG bank #{0}... ", bank);
                dumper.WriteCpu(0xA000, (byte)bank);
                data.AddRange(dumper.ReadCpu(0x8000, 0x4000));
                Console.WriteLine("OK");
            }
        }

        public void DumpChr(FamicomDumperConnection dumper, List<byte> data, int size)
        {
            int banks = size / 0x1000;
            
            for (int bank = 0; bank < banks; bank++)
            {
                Console.Write("Reading CHR bank #{0}... ", bank, bank);
                dumper.WriteCpu(0xB000, (byte)bank); // CHR ROM $FD/0000 bank select
                dumper.WriteCpu(0xC000, (byte)bank); // CHR ROM $FE/0000 bank select
                data.AddRange(dumper.ReadPpu(0x0000, 0x1000));
                Console.WriteLine("OK");
            }
        }

        public void EnablePrgRam(FamicomDumperConnection dumper)
        {
            Console.WriteLine("Warning: SRAM is not supported by this mapper");
        }
    }
}
